web/packages/teleterm/src/ui/services/tshdNotifications/tshdNotificationService.ts

```diff
@@ -83,19 +84,49 @@ export class TshdNotificationsService {
         if (!notificationRequestOneOfIsCannotProxyVnetConnection(subject)) {
           return;
         }
-        const { routeToApp, targetUri, error } =
+        const { routeToApp, targetUri, reason } =
           subject.cannotProxyVnetConnection;
-        const clusterName = routing.parseClusterName(targetUri);
+        const clusterName = this.getClusterName(targetUri);
 
-        return {
-          title: `Cannot connect to ${publicAddrWithTargetPort(routeToApp)}`,
-          description: `A connection attempt to the app in the cluster ${clusterName} failed due to an unexpected error: ${error}`,
-        };
+        switch (reason.oneofKind) {
+          case 'certReissueError': {
+            if (!cannotProxyVnetConnectionReasonIsCertReissueError(reason)) {
+              return;
+            }
+            const { error } = reason.certReissueError;
+
+            return {
+              title: `Cannot connect to ${publicAddrWithTargetPort(routeToApp)}`,
+              description: `A connection attempt to the app in the cluster ${clusterName} failed due to an unexpected error: ${error}`,
+            };
+          }
+          case 'invalidLocalPort': {
+            return {
+              title: `Invalid local port for ${publicAddrWithTargetPort(routeToApp)}`,
+              description: `The port ${routeToApp.targetPort} is not included in target ports of the app ${routeToApp.clusterName} in the cluster ${clusterName}.`,
+              // 3rd-party clients can potentially make dozens of attempts to connect to an invalid
+              // port within a short time. As all notifications from this service go as errors, we
+              // don't want to force the user to manually close each notification.
+              isAutoRemovable: true,
```

Comment:
do we rate-limit or debounce these errors at all, or should we? How are these displayed, do they all pop up as toasts?

Reply:
We don't limit or debounce these in any way, other than the fact that if you get too many of them, the most recent ones will just go off the screen. They do pop up as toasts (like in the screenshot in the PR description) and the RPC they're sent through doesn't cause the Electron app to steal focus.

In the long run, I think these should be integrated either with the notifications from the Web UI somehow. That's why I mostly just don't want to spend too much time on them right now, other than making them auto-removable.

Reply:
>In the long run, I think these should be integrated either with the notifications from the Web UI somehow. That's why I mostly just don't want to spend too much time on them right now, other than making them auto-removable.

Yeah, it would be good to have a place where the notifications are stored. Sometimes it's annoying that they are auto-removed.